### DISCUSSION TITLE
Consider change udf function to lambda style

### DISCUSSION THREAD

**Discussion by sundy-li** [databendlabs/databend#3672] [Ideas]
Current UDF style: `CREATE FUNCTION cal = '@0 + @2 * (@4 / @1) - @3';`

What if there is `String` literal inside?

`CREATE FUNCTION cal = 'concat($0,   \'33\' ) ';`

Need escape.

What if we introduce `lambda` functions to traverse `array` or `maps` in the future?

```
SELECT arrayMap(x -> (x + 2), [1, 2, 3]) as res;
```

What's more, this function did not have function propertype.

In rust, `|a, b|  a + b + 3 `.

We can have a similar function like Rust/ClickHouse.

```
CREATE FUNCTION name AS |parameter0, ...| -> expression
```
Or

```
CREATE FUNCTION name AS (parameter0, ...) -> expression
```

What's your opinion? cc @lianghanzhen

**Comment by bohutang** [comment]
Good idea, this will unify our udf and higher order function on usage for users.
By the way, we already have a plan to support `Lambda functions` in Roadmap2021 https://github.com/datafuselabs/databend/issues/746

**Comment by lianghanzhen** [comment]
As we discussed before, the only difference between `ClickHouse` and current implementation is implicit/explicit parameters. But I do agree that the explicit one would be much more intuitive. Here is the plan to achieve that:

1. Fix the escape issue, may be better parse an expression instead of literal string;
2. Support `Lambda function` style;

Please let me know what your thought is.

**Reply by sundy-li** [reply]
It's ok to go on as you suggest.

**Comment by lianghanzhen** [comment]
@sundy-li @BohuTANG Can you help review this MR of first step: https://github.com/datafuselabs/databend/pull/3684

**Reply by lianghanzhen** [reply]
@sundy-li @BohuTANG This MR should cover all steps.

**Reply by bohutang** [reply]
Sure, thank you!
